buffer: adds buffer.prototype.split / join

eljefedelrodeodeljefe opened the pull request:
<!--
Thank you for submitting a pull request to Node.js. Before you submit, please
review below requirements and walk through the checklist. You can 'tick'
a box by using the letter "x": [x].

Run the test suite by invoking: `make -j4 lint test` on linux or
`vcbuild test nosign` on Windows.

If this aims to fix a regression or you’re adding a feature, make sure you also
write a test. Finally – if possible – a benchmark that quantifies your changes.

Finally, read through our contributors guide and make adjustments as necessary:
https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->
##### Checklist

<!-- remove lines that do not apply to you -->
- [x] tests and code linting passes
- [x] a test and/or benchmark is included
- [x] documentation is changed or added
- [x] the commit message follows commit guidelines
##### Affected core subsystem(s)

<!-- provide affected core subsystem(s) (like doc, cluster, crypto, etc) -->

`buffer`
##### Description of change

<!-- provide a description of the change below this comment -->

> **WIP**
> - benchmarks
> - one less copy (had problems persisting the input buffer in `.split()`)
> - general `c++` related review induced changes, style etc.

**Description/ Motivation**
This PR introduces two additions to the `buffer` API, `.split()` and `.join()`. Motivation is too handle buffers directly (say in streams) and having higher efficiency to what would be possible in userland js. This continues the discussion from https://github.com/nodejs/node/issues/5683 whereas as a `.replace()` API seemed too high level, but can be easily implemented with the methods here. 

**Comparison**
In string scenarios the methods are 1.5x to 3x times slower, whereas a similar JS implementation would be 10x to 20x slower. Detailed benchmarks are yet to be written and also the currently implementation makes at least one copy too much (WIP).

**Usage**
The potential usage is bot only string manipulation, but having a general byte manipulation interface with efficient copying. 

 **For docs:**
##### Class Method: Buffer.split([separator])
- `separator` {Buffer} A separator that gets matched against the input buffer
- Return: {Array} Array of Buffers

`Buffer.split()` behaves similar to `String.prototype.split`, as such it matches
an input against a source buffer, splits the source buffer, removes the input
and returns the number of resulting buffers in an array.

``` js
const buf = new Buffer('Hello World');
const res = buf.split(new Buffer(' '))

console.log(res);
// [ <Buffer 48 65 6c 6c 6f>, <Buffer 57 6f 72 6c 64> ]
```

It is also possible to split with an empty buffer, which results in an array
with the single bytes of buffer contents wrapped in new buffers.

``` js
const buf = new Buffer('Hello');
const res = buf.split(new Buffer(''))

console.log(res);
// [ <Buffer 48>, <Buffer 65>, <Buffer 6c>, <Buffer 6c>, <Buffer 6f> ]
```
##### Class Method: Buffer.join(array[, insert])
- `array` {Array} Array of buffers that will be joined together
- `insert` {Buffer} Buffer that gets inserted between the subjects
- Return: {Buffer}

`Buffer.join()` behaves similar to `Array.prototype.join`, as such it matches
takes an array of buffers and generate new buffer with the concatenated buffers.
Optionally it takes a an `insert` as argument that will be inserted between the
buffers.

``` js
const res = Buffer.join([
  new Buffer('Hello'),
  new Buffer('World'),
], new Buffer('\n'));

console.log(res);
// <Buffer 48 65 6c 6c 6f 0a 57 6f 72 6c 64>
console.log(res.toString());
// Hello
// World
```

_Note: `Array.prototype.join` would fill in commas if no insert will be
provided. `Buffer.join()` though will simply concat against the next buffer._

``` js
const res = Buffer.join([
  new Buffer('Hello'),
  new Buffer('World'),
]);

console.log(res.toString());
// HelloWorld
```

mscdex commented on src/node_buffer.cc:
```diff
@@ -1142,6 +1142,143 @@ void IndexOfNumber(const FunctionCallbackInfo<Value>& args) {
+    }
+  }
+
+  res_buffers = v8::Array::New(env->isolate(), index_arr.size());
```
Would calling out to JS to create this array be faster? I know @trevnorris has done something like this before, but I'm not sure of the circumstances that benefit from it.

eljefedelrodeodeljefe replied:
I guess there are a couple of such optimisations. Happy for input...

trevnorris replied:
If you want to take advantage of this it's `env->push_values_to_array_function()`. Can see the usage of it in `node::GetActiveRequests()` in `src/node.cc`.

addaleax commented on lib/buffer.js:
```diff
@@ -637,6 +637,28 @@ Buffer.prototype.includes = function includes(val, byteOffset, encoding) {
+  if (!limit)
+    limit = null;
+  if (!sep)
+    return [this];
```
I think the docs should note that in this case the resulting `Buffer` is not a copy, as in all other situations? Alternatively: Could maybe all returned `Buffer`s be slices/subarrays? That might be more performant, and users could still create copies if they want to.

trevnorris replied:
very very important that this does return a copy. had a kerfuffle with the `.concat()` method doing the same thing.

trevnorris replied:
This should simply be `if (sep === undefined)`. Allow everything else to be passed to C++, checked and thrown if necessary. We don't want strange values like `NaN` not causing this to throw but `true` to throw.

addaleax commented on src/node_buffer.cc:
```diff
@@ -1142,6 +1142,143 @@ void IndexOfNumber(const FunctionCallbackInfo<Value>& args) {
+  // minus last insert to match Array.prototype.join behavior
+  if (has_insert) {
+    SPREAD_ARG(args[2], insert);
+    buffers_size += num_buffers * insert_length - insert_length; // no last ins
```
`buffers_size` can overflow here and in the above loop

addaleax commented on src/node_buffer.cc:
```diff
@@ -1142,6 +1142,143 @@ void IndexOfNumber(const FunctionCallbackInfo<Value>& args) {
+    memcpy(ins_data, insert_data, insert_length);
+  }
+
+  data = (char*)malloc(buffers_size * sizeof(char));
```
* `sizeof(char)` is defined to be 1 – you can still do this if you think it helps readability, but I’m not sure it’s worth it
* `static_cast<char*>(…)` is preferred to C-style casts in the codebase
* It might be helpful to check that `malloc()` does not return `nullptr` (`CHECK` should suffice, but without it, users might just see `Segmentation fault` when it fails).

addaleax commented on src/node_buffer.cc:
```diff
@@ -1142,6 +1142,143 @@ void IndexOfNumber(const FunctionCallbackInfo<Value>& args) {
+    buffers_size += num_buffers * insert_length - insert_length; // no last ins
+    ins_length = insert_length;
+    // allocate and copy the insert for use in second loop
+    ins_data = (char*)malloc(buffers_size * sizeof(char));
```
Why the copy? Is there any reason not assign `insert_data` and `insert_length` directly to `ins_data` and `ins_length`?

addaleax replied:
Plus this misses the corresponding `free()` right now.

addaleax commented:
I like the additions to the API, but maybe this would be better off in pure JS? I think you can get pretty far using `buffer.indexOf()`, `buffer.subarray()` and `buffer.set()`.

I haven’t dug into the `split()` code too far, but I think there’s something nondeterministic going on…

```
> new Buffer('Hello, World!').split(new Buffer('l'))
[ <Buffer 48 65 6c 6c 6f 2c 20 57>, <Buffer 72 6c 64 21> ]
> new Buffer('Hello, World!').split(new Buffer('l'))
[ <Buffer 48 65 6c 6c 6f 2c 20 57>, <Buffer 72 6c 64 21> ]
> new Buffer('Hello, World!').split(new Buffer('l'))
[ <Buffer 48 65 6c 6c>,
  <Buffer >,
  <Buffer >,
  <Buffer >,
  <Buffer >,
  <Buffer 72 6c 64>,
  <Buffer > ]
> new Buffer('Hello, World!').split(new Buffer('l'))
[ <Buffer 48 65 6c 6c 6f 2c 20 57>, <Buffer 72 6c 64 21> ]
```

trevnorris commented:
I would like to see benchmark comparisons between a fully native implementation of `split()` vs using `indexOf()`/`copy()`.

trevnorris commented on lib/buffer.js:
```diff
@@ -637,6 +637,28 @@ Buffer.prototype.includes = function includes(val, byteOffset, encoding) {
+  if (!sep)
+    return [this];
+  if (!Buffer.isBuffer(sep))
+    throw new TypeError('First argument must be a buffer.');
```
This is useless. Always use `Buffer::HasInstance()` in C++ and throw from there. The proto here can be screwed with.

trevnorris commented on lib/buffer.js:
```diff
@@ -637,6 +637,28 @@ Buffer.prototype.includes = function includes(val, byteOffset, encoding) {
+// Limit currently not implemented; reserved
+Buffer.prototype.split = function split(sep, limit) {
+  if (!limit)
+    limit = null;
```
no need for this check. pass it along and allow c++ to throw if needed.

trevnorris replied:
sorry. meant this for `if (!sep)` below.

trevnorris commented on lib/buffer.js:
```diff
@@ -637,6 +637,28 @@ Buffer.prototype.includes = function includes(val, byteOffset, encoding) {
+Buffer.join = function join(array, insert) {
+  if (!insert)
+    insert = null;
+  if (!Array.isArray(array) || !Buffer.isBuffer(array[0]))
```
Do both checks in C++ and throw from there. No need to check `insert` here. Check using `HasInstance()` in C++ and throw from there.

trevnorris commented on lib/buffer.js:
```diff
@@ -637,6 +637,28 @@ Buffer.prototype.includes = function includes(val, byteOffset, encoding) {
+    throw new TypeError('First argument must be an array of buffers with ' +
+                                                        'at least one buffer.');
+
+  return binding.join(array, array.length, insert);
```
the length can be easily retrieved from C++. no need to pass it from here.

trevnorris commented on src/node_buffer.cc:
```diff
@@ -1142,6 +1142,143 @@ void IndexOfNumber(const FunctionCallbackInfo<Value>& args) {
+void Split(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  THROW_AND_RETURN_UNLESS_BUFFER(env, args[0]);
```
this should be set for `args[1]` as well.

trevnorris commented on src/node_buffer.cc:
```diff
@@ -1142,6 +1142,143 @@ void IndexOfNumber(const FunctionCallbackInfo<Value>& args) {
+
+  std::vector<size_t> index_arr;
+  Local<Object> obj;
+  Persistent<Object> persistent(env->isolate(), obj);
```
i'm missing why this is necessary.

trevnorris commented on src/node_buffer.cc:
```diff
@@ -1142,6 +1142,143 @@ void IndexOfNumber(const FunctionCallbackInfo<Value>& args) {
+  std::vector<size_t> index_arr;
+  Local<Object> obj;
+  Persistent<Object> persistent(env->isolate(), obj);
+  Local<v8::Array> res_buffers;
```
if it's not being used yet then put `using v8::Array` at the top of file.

trevnorris commented on src/node_buffer.cc:
```diff
@@ -1142,6 +1142,143 @@ void IndexOfNumber(const FunctionCallbackInfo<Value>& args) {
+  // copy bytes from index bounds to new array of buffers
+
+  // REVIEW(eljefedelrodeodeljefe): copy first necessary?
+  char* buf = new char[js_buf_length];
```
use `malloc()`, check if `nullptr`. if so then throw an error that data couldn't be allocated. this will abort if that's the case.

trevnorris commented on src/node_buffer.cc:
```diff
@@ -1142,6 +1142,143 @@ void IndexOfNumber(const FunctionCallbackInfo<Value>& args) {
+  // Intentional no-op
+}
+
+void Split(const FunctionCallbackInfo<Value>& args) {
```
This should be leveraging a lot of what's in `Buffer::IndexOfBuffer()`. Would simply this and keep implementations unified.

Fishrock123 commented:
There's been a bit of a trend recently... Do we really need to add the kitchen sink (everything) to buffer?

rvagg commented:
I share @Fishrock123's concerns about `Buffer`, particularly since a changes being made and proposed are introducing new APIs purely for the sake of performance, when the same functionality can be achieved with existing APIs. While there might be a good case for introducing new APIs for common cases, the need to do so suggests we may need to step back and take a more high-level view of `Buffer` as a whole and be more strategic in its API rather than just letting it expand ad-hoc. Perhaps there are broader shifts we can make with the API to address concerns coming up. We also have to consider the overlap we have with the `TypedArray` API that we now have, we're not just dealing with what hangs directly off `Buffer` anymore and we have to expect TC39 to throw more stuff into `TypedArray` and they don't exactly have an awesome track record for paving cowpaths so I'd expect conflicts to come up.

mhdawson commented:
I'll add my 2  cents that I would prefer planned versus ad-hoc expansion.  I'd also ask if we have end-user uses cases other than micro-benchmarks that are slower for the lack of these methods ?  For example if these are added are there specific applications or frameworks that we know will use them and benefit ?

jasnell commented:
I'm not so keen on `split()` and `join()` can be fairly easily handled by the existing `concat`. 

As a use case, we can consider the new `Buffer` support in the `fs` module. On Linux, it's possible (albeit rare) for a file path to actually contain multiple encodings even within a single path. For instance, you could have the base path `/usr` as ASCII bytes, a UTF-8 encoded home dir, and an ISO-8859-1 encoded directory, all of which needs to be joined by the path separator.

```js
const res = Buffer.join([
  Buffer.alloc(0), // to ensure we get a leading path.sep
  Buffer.from('usr', 'ascii'),
  Buffer.from('joë', 'utf8'),
  Buffer.from('t\xe9st', 'binary')
], Buffer.from(path.sep));
```

The equivalent method using the existing concat API is:

```js
const sep = Buffer.from(path.sep);
const base = Buffer.from('usr', 'ascii');
const user = Buffer.from('joë', 'utf8');
const folder = Buffer.from('t\xe9st', 'binary');
Buffer.concat([sep, base, sep, user, sep, folder], 
                      base.length + user.length + folder.length + (sep.length * 3));
```

In this case, `concat` works just fine.

trevnorris commented:
@jasnell I'm not sure what you're referring to. Here's an equivalent implementation of `join()` in JS that has the same documented functionality:
```js
function join(array, insert) {
  // Buffer.concat() does isArray() check. No need for it here.
  if (insert === undefined)
    return Buffer.concat(array);
  if (!ArrayBuffer.isView(insert) || !(insert instanceof Buffer))
    throw new TypeError('insert must be a Buffer');
  const narray = new Array(array.length * 2 - 1);
  for (var i = 0; i < narray.length; i++)
    narray[i] = (i % 2) === 0 ? array[i >> 1] : insert;
  return Buffer.concat(narray);
}

console.log(join([
    Buffer(0),
    Buffer.from('usr', 'ascii'),
    Buffer.from('joë', 'utf8'),
    Buffer.from('t\xe9st', 'binary')], Buffer('/')).toString());
// output: /usr/joë/t�st
```
I think the point is this easily can live outside of core.

jasnell commented:
Sorry, yes, the point I was attempting to make is that this is something that is easily doable with the existing APIs

trevnorris commented:
@jasnell My bad. Read "can be fairly" as "can't be fairly".

eljefedelrodeodeljefe commented:
hmm. I don't know about `Buffer` being a sink. But to me it's kind of natural when one defines a data type. Also to provide a comprehensive API then.

While I agree on `.join()` being a little redundant to `.concat()`, I absolutely cannot re `.split()`. Due to a number of reasons.

> **TL;DR** I regard `Buffer` as Node's way of manipulating data on the `char` / `byte` / `8bit` level. In that sense the API is missing an effective scissor. Please have a look at `golang`'s <code>[**bytes**](https://golang.org/pkg/bytes/#Index)</code>

1. recursive `.indexOf()` is incredibly slow, not only in our implementation, but in general as API. I regard it as a flaw in (nearly every) language to not provide `.indexes() => []` from the beginning, if there is no decent `.split()` API. This is because recursive `.indexOf()` in adds **a lot** boilerplate and head wrapping, as you can see in my `js` implementation in the issue. (I will add benchmarks as next step.). 
<br><br>
Let alone that potentially jump in and out of c++ with the various type checks potentially adds up (tossing numbers) a 10x performance degradation, which I wouldn't call micro-benchmark, when e.g. applied streams. <br>
<br>
2. `.slice()`would require you to have that indexing and moving cursors yourself. Which might be an easy enough task, but a very repetitive one, that leaves you wondering why this hasn't just been build in. <br> <br>
3. `TypedArray` has no built-in `.split()` method afaik. My interpretation is, that they haven't regarded it as necessary, because they haven't concepted it as `Buffer` data type.<br><br>
4. On the competition side of things we can see that we are still far off from the width of the `golang` `bytes` API, see [here](https://golang.org/pkg/bytes/), yet in light of data manipulation it makes perfectly sense. Data manipulation: (don't wanna lecture here and unfortunately haven't found resources on the topic) it's important to provide basic API, being **create**, **delete**, **update**, **read**, **copy**.  Whereas the current API doesn't _really_ have an effective update mechanism. The reason why `.slice()` doesn't work well is it being concepted for, what the name suggests, cutting a smaller piece from a bigger piece and do some work on the smaller piece. `.split()` on the other hand has the larger amount of data in mind and `.join()`moving the pieces together to fulfill the promise of updating.<br><br>

jasnell commented:
If we do add `Buffer.prototype.split()`, I would very much like to consider simpler argument forms *in addition* to passing in a `Buffer` as the separator. For instance, it ought to be possible to do:

```js
const buf = getBufferSomehow();

buf.split(sepBuf); // where sepBuf is the sep as a buffer, limit is the maximum number of splits
buf.split(singleByte); // where singleByte is a single octet used as the split delim
buf.split(sep, encoding); // where sep is a string and encoding is an optional char encoding
buf.split(array); // where array is a javascript array of octets used as the split delim
```

Also, I see that `limit` is there but not currently being used. I definitely think we'd want `limit` to be implemented from the start, and while it would add some additional complexity, I'm thinking that we'd likely want a starting offset and ending offset as well, e.g.

```js
buf.split(0xFF, 2, 0, 100); // equivalent to buf.slice(0, 100).split(0xFF, 2)
```

eljefedelrodeodeljefe commented:
@jasnell I think those additions are really reasonable and quite nice. Happy to employ those if general notion towards `.split()` is leaning positive. I'll wait another day.

```js
buf.split(array); // as in: matches _n_ octets, I assume. Opposite'd be: matches consecutive octets
```

Thx.

jasnell commented:
For `buf.splif(array)` it would be the equivalent to `buf.split(Buffer.from(array))`

eljefedelrodeodeljefe commented:
So, I profiled this quite heavily and came to the realization, that this should stay implemented in `js`. 

## Findings

* direct "pointer pushing" might be fast, but v8 implements it equally as well <br><br>
* handling constructors via setting prototypes of the new Buffers - in order not to construct totally new objects  in `c++` - might be okay, but feels very dirty. Also this is nicely and cleanly implemented in `.slice()`<br><br>
* speaking of `.slice()`: since we know `.subarray()` is very effective due to handing back buffer views, I would have wanted to implement this in `c++`, however I couldn't find the v8 implementation and was discouraged by the above to continue.<br><br>
* indexing might be faster in `c++`, but `.indexOf()`'s search algorithm does not pay a too high extra effort in the rather simple byte search case here. Since crossing the boundary and constructing a `js` array invokes some cost and I couldn't profile it very well, I switched back to `js`. In any case the gain is probably to low.

## Result

1. So let alone for maintainability and easiness to alter the API I am keeping this in `js`. Future improvements could be thoroughly investigated (also in `c++`) in multiple directions, i.e. string decoding the buffer and and work with native string functions, which score quite well in the benchmarks. However I wouldn't do this, since you'll, for example, would need to set prototypes after you have created arrays.<br><br>
2. The API now allows @jasnell inputs and limit. Minus:
  - I was holding back start to end though because it felt convoluted, and could be implemented by slicing the buffer beforehand, I guessed.
  - single octets outside of an array are not allowed by the constructor API, afaik. So I omitted this in order not to have an extra case to check for<br><br>
3. Benchmarks:
  -- the `cpp` solution w/o the optimizations I discussed above was 50% as fast as the `js` one presented here. I abandoned `c++` for exactly that reason.
  -- as a counter example now serves a simple `string.split().join()`, which doesn't have the the constructor building within `.split()`.
  -- if benchmarks are correct this results in the Buffer implementation be 64% as fast as the String one for big Buffers (100mb). <br><br> 
  ```console
buffers/buffer-split.js size=small assert=true type=js_native n=1024: 244.65256
buffers/buffer-split.js size=small assert=true type=string n=1024: 700.58533
buffers/buffer-split.js size=small assert=false type=js_native n=1024: 275.21173
buffers/buffer-split.js size=small assert=false type=string n=1024: 820.37397
buffers/buffer-split.js size=big assert=true type=js_native n=1024: 188.80848
buffers/buffer-split.js size=big assert=true type=string n=1024: 262.13276
buffers/buffer-split.js size=big assert=false type=js_native n=1024: 191.48025
buffers/buffer-split.js size=big assert=false type=string n=1024: 295.85797
```
4. Concerning `.replace()`: 
It turns out that in a replace-scenario the fastest implementation would rather be a similar one I presented in the old issue. If this is a go ahead and the wider Buffer API is generally welcomed then I would suggest including `.replace()` also, because it's very easy to do.

5. I have externalized a helper function: `.indexes()`. The abstracts the while loop for `.indexOf()` and returning the array. It could be considered to expose this via documentation - I am arguing for it as I did before. I find languages lacking this to catch simple QA stuff like "<favorite lang>: find all occurences in <any construct>"

## Go-ahead
Apart from the implementation, how would be the direction of settling this?
Be my guest for code review.

addaleax commented on lib/buffer.js:
```diff
@@ -658,6 +658,87 @@ Buffer.prototype.includes = function includes(val, byteOffset, encoding) {
+Buffer.join = function join(array, insert) {
+  if (array.length === 0)
+    throw new TypeError('first argument must be an array of Buffers');
+  if (array.length === 1)
```
`if (array.length === 1 || !insert)`? That would save you basically all the case distinctions below, I think.

addaleax commented on lib/buffer.js:
```diff
@@ -658,6 +658,87 @@ Buffer.prototype.includes = function includes(val, byteOffset, encoding) {
+    // push from index to index; catch last case after loop
+    for (let i = 0; i < len; i++) {
+      res.push(this.slice(cursor, indexes[i]));
+      cursor += indexes[i] - cursor + sepLen;
```
I think the whole point of this is to have compatibility with `String.prototype.split`, right? I think you have to handle the case where the `sep` strings overlap differently, then:
```js
> 'aaabaaa'.split('aa')
[ '', 'ab', 'a' ]
> Buffer.from('aaabaaa').split(Buffer.from('aa'))
[ <Buffer >, <Buffer >, <Buffer 62>, <Buffer >, <Buffer > ]
```

trevnorris commented on lib/buffer.js:
```diff
@@ -658,6 +658,87 @@ Buffer.prototype.includes = function includes(val, byteOffset, encoding) {
+  if (!limit && typeof encoding === 'number')
+    limit = encoding;
+  if (sep)
+    sep = Buffer.from(sep, encoding); // allows all constructor APIs
```
unnecessary overhead. if `sep` is a buffer then `Buffer.from()` will make a copy of it. since we're not changing the values of `sep` there's no need for this.

trevnorris commented on lib/buffer.js:
```diff
@@ -658,6 +658,87 @@ Buffer.prototype.includes = function includes(val, byteOffset, encoding) {
   return this.indexOf(val, byteOffset, encoding) !== -1;
 };
 
+Buffer.prototype.indexes = function indexes(match, limit) {
```
what's the utility for having this exposed? it's not even documented.

trevnorris commented on doc/api/buffer.md:
```diff
@@ -589,6 +589,71 @@ console.log(bufA.length);
+
+`Buffer.split()` behaves similar to `String.prototype.split`, as such it matches
+an input against a source buffer, splits the source buffer, removes the input
+and returns the number of resulting buffers in an array.
```
This never specifies whether it'll return a slice or a copy. Should be defined, and make sure the implementation matches. I don't care much either way, but copy may be more intuitive.

trevnorris commented on lib/buffer.js:
```diff
@@ -658,6 +658,87 @@ Buffer.prototype.includes = function includes(val, byteOffset, encoding) {
+};
+
+// not overriding prototype here, because of Uint8Array, REVIEW
+Buffer.join = function join(array, insert) {
```
We happily override Uint8Array methods wherever is convenient. Just make sure it's a superset of functionality, so the Uint8Array method still executes as expected. This should be on the prototype.

trevnorris commented on src/node_buffer.cc:
```diff
@@ -1226,6 +1226,7 @@ void Initialize(Local<Object> target,
   env->SetMethod(target, "readFloatBE", ReadFloatBE);
   env->SetMethod(target, "readFloatLE", ReadFloatLE);
 
+
```
unnecessary whitespace change.

jasnell commented:
@eljefedelrodeodeljefe ... still want to pursue this?

eljefedelrodeodeljefe commented:
lack of interest -> rather not
